File: src/main/ATOMValue.java
package main;

import java.util.List;

public class ATOMValue extends ATOMElement {

    private ATOMValueType type;

    private boolean boolVal;
    private int intVal;
    private String strVal;
    private SparseArrayList<ATOMValue> arrVal;
    private ATOMScope objVal;

    public ATOMValue() {
    }

    public ATOMValue(boolean val) {
        type = ATOMValueType.BOOLEAN;
        boolVal = val;
    }

    public ATOMValue(int val) {
        type = ATOMValueType.INT;
        intVal = val;
    }

    public ATOMValue(String val) {
        type = ATOMValueType.STRING;
        strVal = val;
    }

    public ATOMValue(List<ATOMValue> val) {
        type = ATOMValueType.ARRAY;
        arrVal = new SparseArrayList<>(val);
    }

    public ATOMValue(ATOMScope val) {
        type = ATOMValueType.OBJECT;
        objVal = val;
    }

    public static ATOMValue NULL() {
        ATOMValue newVal = new ATOMValue();
        newVal.type = ATOMValueType.NULL;
        return newVal;
    }

    public ATOMValue eval() {
        return this;
    }

    public ATOMValue compute() {
        if (type == ATOMValueType.OBJECT) {
            return objVal.compute();
        }
        return this;
    }

    public String toString() {
        switch (type) {
            case OBJECT:
                return objVal.toString();
            case ARRAY:
                return arrVal.toString();
            case INT:
                return "" + intVal;
            case STRING:
                return '"' + strVal + '"';
            case BOOLEAN:
                return "" + boolVal;
            case NULL:
                return "NULL";
            default:
                return null;
        }
    }

    public boolean isTruthy() {
        switch (type) {
            case NULL:
                return false;
            case BOOLEAN:
                return boolVal;
            case INT:
                return intVal != 0;
            case ARRAY:
                return true;
            case OBJECT:
                return true;
        }
        return true;
    }

    public ATOMValueType getType() {
        return type;
    }

    protected void setType(ATOMValueType newType) {
        type = newType;
        if (newType == ATOMValueType.ARRAY) {
            arrVal = new SparseArrayList<>();
        }
    }

    public boolean getBoolVal() {
        return boolVal;
    }

    public int getIntVal() {
        return intVal;
    }

    public String getStrVal() {
        return strVal;
    }

    public SparseArrayList<ATOMValue> getArrVal() {
        return arrVal;
    }

    public ATOMScope getObjVal() {
        return objVal;
    }

    public boolean equals(ATOMValue other) {
        switch (this.type) {
            case NULL:
                return true;
            case BOOLEAN:
                return this.boolVal == other.boolVal;
            case INT:
                return this.intVal == other.intVal;
            case STRING:
                return this.strVal.equals(other.strVal);
        }
        return this == other;
    }

    public static boolean listEquals(List<ATOMValue> first, List<ATOMValue> second) {
        if (first.size()!=second.size()) {
            return false;
        }
        for (int i=0;i<first.size();i++) {
            if (!first.get(i).equals(second.get(i))) {
                return false;
            }
        }
        return true;
    }

}

File: web_interpreter/atom/ATOMValue.js
import SparseArrayList from './SparseArrayList.js';

export default class ATOMValue {

  constructor(val) {
    this.type = null;
    this.boolVal = null;
    this.intVal = null;
    this.strVal = null;
    this.arrVal = null;
    this.objVal = null;
    if (typeof val == 'boolean') {
      this.type = 'BOOLEAN';
      this.boolVal = val;
    }
    if (typeof val == 'number') {
      this.type = 'INT';
      this.intVal = val;
    }
    if (typeof val == 'string') {
      this.type = 'STRING';
      this.strVal = val;
    }
    if (typeof val == 'object') {
      if (Array.isArray(val)) {
        this.type = 'ARRAY';
        this.arrVal = val;
      } else {
        this.type = 'OBJECT';
        this.objVal = val;
      }
    }
  }

  static NULL() {
    let toReturn = new ATOMValue();
    toReturn.type = 'NULL';
    return toReturn;
  }

  eval() {
    return this;
  }

  compute() {
    if (this.type == 'OBJECT') {
      return this.objVal.compute();
    }
    return this;
  }

  toString() {
    switch(this.type) {
      case 'OBJECT':
        return this.objVal.toString();
      case 'ARRAY':
        return '['+this.arrVal.toString()+']';
      case 'INT':
        return '' + this.intVal;
      case 'STRING':
        return '"' + this.strVal + '"';
      case 'BOOLEAN':
        return '' + this.boolVal;
      case 'NULL':
        return "NULL";
      default:
        return null;
    }
  }

  isTruthy() {
    switch(this.type) {
      case 'NULL':
        return false;
      case 'BOOLEAN':
        return this.boolVal;
      case 'INT':
        return this.intVal != 0;
      case 'ARRAY':
        return true;
      case 'OBJECT':
        return true;
    }
    return true;
  }

  getType() {
    return this.type;
  }

  setType(newType) {
    this.type = newType;
    if (newType == 'ARRAY') {
      arrVal = new SparseArrayList();
    }
  }

  equals(other) {
    switch(this.type) {
      case 'NULL': return true;
      case 'BOOLEAN': return this.boolVal == other.boolVal;
      case 'INT': return this.intVal == other.intVal;
      case 'STRING': return this.strVal == other.strVal;
    }
    return this==other;
  }

  static listEquals(first, second) {
    if (first.length!=second.length) {
      return false;
    }
    for (let i=0;i<first.length;i++) {
      if (!first[i].equals(second[i])) {
        return false;
      }
    }
    return true;
  }

}